File: packages/reader_client/reader_client/lib/src/models/tags/tag_scan_data.dart
import 'dart:ffi';

import 'package:equatable/equatable.dart';
import 'package:reader_client/reader_client.dart';

class TagScanData extends Equatable{
  const TagScanData({
    this.epc,
    this.tidData,
    this.rssi,
    this.pc,
    this.crc,
    this.channelFrequency,
  });

  final String? epc;
  final String? tidData;
  final int? rssi;
  final int? pc;
  final int? crc;
  final int? channelFrequency;

  factory TagScanData.fromJson (Map<String, dynamic> data) {
    return TagScanData(
        epc: data['epc'] as String,
        tidData: data['tidData'] as String?,
        rssi: data['rssi'] as int?,
        pc: data['pc'] as int?,
        crc: data['crc'] as int?,
        channelFrequency: data['channelFrequency'] as int?,
    );
  }
  TagScanData copyWith({
    String? epc,
    String? tidData,
    int? rssi,
    int? pc,
    int? crc,
    int? channelFrequency,
}){
    return TagScanData(
      epc: epc ?? this.epc,
      tidData: tidData ?? this.tidData,
      rssi: rssi ?? this.rssi,
      pc: pc ?? this.pc,
      crc: crc ?? this.crc,
      channelFrequency: channelFrequency ?? this.channelFrequency,
    );
  }

  @override
  List<Object?> get props => [
    epc,
    tidData,
    rssi,
    pc,
    crc,
    channelFrequency,
  ];
}

File: android/app/src/main/kotlin/com/mtg/zimble/reader/tags/data/TagScanData.kt
package com.mtg.zimble.reader.tags.data

import com.google.gson.Gson

class TagScanData (
    var epc: String,
    var tidData: String?,
    var rssi: Int?,
    var pc: Int?,
    var crc: Int?,
    var channelFrequency: Int?,
){
    /// Function to build a MessageMap from the TagData object to send across the EventChannel
    fun toMessageMap(): MutableMap<String, Any?> {
        var map = mutableMapOf<String, Any?>()
        var returnMap = mutableMapOf<String, Any?>()
        var tagDataMap = mutableMapOf<String, Any?>(
            "epc" to epc,
            "tid" to tidData,
            "rssi" to rssi,
            "pc" to pc,
            "crc" to crc,
            "channelFrequency" to channelFrequency,
        )
        map["tagScanData"] = Gson().toJson(tagDataMap)
        return returnMap
    }

    /// Function to build a Json String from the TagData object
    ///  to send across the EventChannel or to remote API
    fun toJson(): String {
        var returnMap = mutableMapOf<String, Any?>()
        //var returnMap = mutableMapOf<String, Any?>()

        returnMap["tagScanData"] = mutableMapOf<String, Any?>(
            "epc" to epc,
            "tid" to tidData,
            "rssi" to rssi,
            "pc" to pc,
            "crc" to crc,
            "channelFrequency" to channelFrequency,
        )

        //returnMap.put("tagData", map)
        return Gson().toJson(returnMap)
    }

    companion object {
        /// Function to build an Initial TagData object for the Stateflow (since it cant have
        /// a null starting value and will be EPC = INITIAL and ignored by the collector
        fun  initial(): TagScanData {
            return TagScanData(
                epc = "INITIAL",
                tidData = null,
                rssi = null,
                pc = null,
                crc = null,
                channelFrequency = null,
            )
        }


        /// Function to build the TagData object from a MutableMap
        /// (usually received from the reader in the AndroidTagController)
        fun fromMap(tagDataMap: MutableMap<String, Any?>): TagScanData {
            return TagScanData(
                epc = tagDataMap["epc"].toString(),
                tidData = tagDataMap["tidData"] as String?,
                rssi = tagDataMap["rssi"] as Int?,
                pc = tagDataMap["pc"] as Int?,
                crc = tagDataMap["crc"] as Int?,
                channelFrequency = tagDataMap["channelFrequency"] as Int?,
            )
        }
    }
}